Sources/KituraSession/SessionState.swift

```diff
@@ -110,4 +110,118 @@ public class SessionState {
             isDirty = true
         }
     }
+    
+    // Check if the provided value is a primative JSON type.
+    private func isPrimative(value: Any) -> Bool {
+        if value is [Any] {
+            return value is [String] ||
+                value is [Int] ||
+                value is [Double] ||
+                value is [Bool]
+        } else {
+            return value is String ||
+                value is Int ||
+                value is Double ||
+                value is Bool
+        }
+    }
+    // MARK: Codable session
+    
+    /**
+     Encode an encodable value as JSON and store it in the session for the provided key.
+     ### Usage Example: ###
+     The example below defines a `User` struct.
+     It decodes a `User` instance from the request body
+     and stores it in the request session using the user's id as the key.
+     ```swift
+     public struct User: Codable {
+        let id: String
+        let name: String
+     }
+     let router = Router()
+     router.all(middleware: Session(secret: "secret"))
+     router.post("/user") { request, response, next in
+         let user = try request.read(as: User.self)
+         try request.session?.add(user, forKey: user.id)
+         response.status(.created)
+         response.send(user)
+         next()
+     }
+     ```
+     - Parameter value: The Encodable object which will be added to the session.
+     - Parameter forKey: The key that the Encodable object will be stored under.
+     - Throws: `EncodingError` if value to be stored fails to be encoded as JSON.
+     - Throws: `SessionCodingError.failedToSerializeJSON` if value to be stored fails to be serialized as JSON.
+     */
+    public func add<T: Encodable>(_ value: T, forKey key: String) throws {
+        let json: Any
+        if isPrimative(value: value) {
+            json = value
```

Comment:
Why do we need this special casing for primitives?

Reply:
JSONEncoder and JSONDecoder do not have an .allowFragments option and will throw an error if you try and encode/decode a primitive type using them. This has been raised as a bug in [SR-6163](https://bugs.swift.org/browse/SR-6163).
Until that is implemented, isPrimitive handles the case where someone wants to store/retrieve a primitive type or array of primitives type

Reply:
OK - I think we should consider not allowing primitives and throwing if the serialization fails. What do you think?